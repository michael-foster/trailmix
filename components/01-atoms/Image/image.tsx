import React from 'react';
import useNativeLazyLoading from '@charlietango/use-native-lazy-loading';
import { useInView } from 'react-intersection-observer';

export type ImageProps = {
  /** Maps to the <img> tag's alt attribute */
  alt: string;
  /** Class names to pass to <img> */
  classes?: string;
  /** Image height */
  height?: string;
  /** Determines if the image is lazy loaded */
  shouldLazyLoad?: boolean;
  /** Maps to the media attribute on the <source> tags within the <picture> tag. The sizes should be comma-separated.  */
  sizes?: string;
  /** Maps to the <img> tag's src attribute */
  src: string;
  /** Maps to the the srcSet attribute on the <source> tags within the <picture> tag. The sources should be comma-separated. */
  srcSet?: string;
  /** Image width */
  width?: string;
  /** whether or not to apply full width to the image */
  fullWidth?: boolean;
};

type PictureProps = {
  testId: string;
};

export const Image = ({
  alt,
  classes = '',
  height = 'auto',
  shouldLazyLoad = true,
  sizes = '',
  src = '',
  srcSet = '',
  width = 'auto',
  fullWidth = true
}: ImageProps) => {
  const supportsLazyLoading = useNativeLazyLoading();
  const { ref, inView } = useInView({
    triggerOnce: true,
    rootMargin: '200px 0px',
    skip: supportsLazyLoading !== false
  });

  const Picture = ({ testId }: PictureProps) => {
    const loadingType = shouldLazyLoad ? 'lazy' : 'eager';
    const sourcesArray = srcSet.split(', ');
    const sizesArray = sizes.split(', ');
    const showImage = inView || supportsLazyLoading || !shouldLazyLoad;
    const imageClasses = `${classes} ${fullWidth ? 'min-w-full' : ''}`.trim();

    if (sourcesArray.length > 1) {
      return (
        <picture data-testid={testId}>
          {sourcesArray.map((source, i) => (
            <source
              /* eslint-disable react/no-array-index-key */
              key={`${source}-${i}`}
              media={sizesArray?.[i] || ''}
              srcSet={source}
            />
          ))}
          {showImage ? (
            <img
              className={imageClasses}
              alt={alt}
              src={src}
              loading={loadingType}
              style={{ height, width }}
            />
          ) : null}
        </picture>
      );
    }

    if (showImage) {
      return (
        <img
          data-testid={testId}
          alt={alt}
          className={imageClasses}
          loading={loadingType}
          src={src}
          srcSet={srcSet}
          style={{ height, width }}
        />
      );
    }
    return null;
  };

  if (shouldLazyLoad) {
    return (
      <div ref={ref}>
        <Picture testId="lazy-loaded-picture" />
      </div>
    );
  }

  return <Picture testId="eager-loaded-picture" />;
};
