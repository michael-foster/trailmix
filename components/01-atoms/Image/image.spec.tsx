import React from 'react';
import { render, screen } from '@testing-library/react';
import { useInView } from 'react-intersection-observer';
import * as useNativeLazyLoading from '@charlietango/use-native-lazy-loading';
import { Image } from './image';

jest.mock('react-intersection-observer');

const mockUseInView = useInView as jest.Mock;

type PropOverrides = {
  shouldLazyLoad?: boolean;
  sizes?: string;
  srcSet?: string;
  fullWidth?: boolean;
};

const setup = (propOverrides: PropOverrides = {}) => {
  const props = {
    alt: 'runner',
    classes: 'cool neat wow',
    src: 'https://www.womensrunning.com/runner.jpg',
    ...propOverrides
  };
  const { asFragment } = render(<Image {...props} />);

  return { asFragment };
};

const mockInView = (isInView: boolean) => {
  mockUseInView.mockImplementation(() => ({
    inView: isInView
  }));
};

const mockUseNativeLazyLoading = (isSupported: boolean) => {
  jest.spyOn(useNativeLazyLoading, 'default').mockReturnValue(isSupported);
};

describe('Image', () => {
  beforeEach(() => {
    mockInView(true);
    mockUseNativeLazyLoading(true);
  });

  test('single source snapshot', () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });

  test('multi-source snapshot', () => {
    const { asFragment } = setup({
      srcSet:
        'http://www.outside.com/canoe-small.jpg, http://www.outside.com/canoe-large.jpg, http://www.outside.com/canoe-large.jpg',
      sizes: '(max-width: 300px), (max-width: 600px), (max-width: 900px)'
    });

    expect(asFragment()).toMatchSnapshot();
  });

  test('optional full width snapshot', () => {
    const { asFragment } = setup({ fullWidth: false });

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders a lazy loaded image when in view and native lazy loading unavailable', () => {
    mockInView(true);
    mockUseNativeLazyLoading(false);

    setup();

    const lazyLoadedPicture = screen.getByTestId('lazy-loaded-picture');

    expect(lazyLoadedPicture).toBeTruthy();
  });

  it('renders a lazy loaded image natively', () => {
    mockInView(false);
    mockUseNativeLazyLoading(true);

    setup();

    const lazyLoadedPicture = screen.getByTestId('lazy-loaded-picture');

    expect(lazyLoadedPicture).toBeTruthy();
  });

  it('renders an eager loaded image', () => {
    setup({ shouldLazyLoad: false });

    const eagerLoadedPicture = screen.getByTestId('eager-loaded-picture');

    expect(eagerLoadedPicture).toBeTruthy();
  });

  it("doesn't render an image if it's not in view and lazy loading isn't supported by the browser", () => {
    mockInView(false);
    mockUseNativeLazyLoading(false);

    setup();

    const image = screen.queryByRole('img');

    expect(image).not.toBeInTheDocument();
  });
});
