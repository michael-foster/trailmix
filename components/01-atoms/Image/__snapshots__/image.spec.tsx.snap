// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image multi-source snapshot 1`] = `
<DocumentFragment>
  <div>
    <picture
      data-testid="lazy-loaded-picture"
    >
      <source
        media="(max-width: 300px)"
        srcset="http://www.outside.com/canoe-small.jpg"
      />
      <source
        media="(max-width: 600px)"
        srcset="http://www.outside.com/canoe-large.jpg"
      />
      <source
        media="(max-width: 900px)"
        srcset="http://www.outside.com/canoe-large.jpg"
      />
      <img
        alt="runner"
        class="cool neat wow min-w-full"
        loading="lazy"
        src="https://www.womensrunning.com/runner.jpg"
        style="height: auto; width: auto;"
      />
    </picture>
  </div>
</DocumentFragment>
`;

exports[`Image optional full width snapshot 1`] = `
<DocumentFragment>
  <div>
    <img
      alt="runner"
      class="cool neat wow"
      data-testid="lazy-loaded-picture"
      loading="lazy"
      src="https://www.womensrunning.com/runner.jpg"
      srcset=""
      style="height: auto; width: auto;"
    />
  </div>
</DocumentFragment>
`;

exports[`Image single source snapshot 1`] = `
<DocumentFragment>
  <div>
    <img
      alt="runner"
      class="cool neat wow min-w-full"
      data-testid="lazy-loaded-picture"
      loading="lazy"
      src="https://www.womensrunning.com/runner.jpg"
      srcset=""
      style="height: auto; width: auto;"
    />
  </div>
</DocumentFragment>
`;
